### PR TITLE
[#2701] Fix OS version checks.

### PR DESCRIPTION
**Pull request by dumol**
Problem?
-----------
OS version checks are broken for Solaris versions older then Solaris 10 and for OS X version 10.10 or newer.

Solution?
-----------
Re-factor OS version checks to only compare integers using the relevant bits in the OS version numbers, no matter how many dot-delimited fields they have. Eg, for RHEL, the OS version numbers may be: `4`, `5.11`, `6.0` or `7.0.1406`, while for AIX they could be `5.3.0.0` or `7.1.0.0`, for Solaris `5.9` or `5.10` and for OS X `10.8` or `10.10.1`.

**Drive-by fixes**:
  * avoid confusing Fedora with RHEL or OpenSUSE with SLES
  * let Ubuntu non-LTS versions take the generic Linux path, as they are not necessarily backwards-compatible with the previous LTS version.

How to check?
-----------------
Please review changes (you may ignore style guide fixes).
Run the tests.
Check the logic of `bash -x paver.sh detect_os` in AIX, Solaris, OS X and supported Linux distros.

reviewer: @adiroiban

### CODE REVIEW

**Comment by dumol** [comment]
**needs-review**

File: paver.sh [hunk 1]
```diff
@@ -28,6 +28,8 @@ set -o pipefail
 # Initialize default value.
 COMMAND=${1-''}
 DEBUG=${DEBUG-0}
+LOCAL_PYTHON_BINARY_DIST=""
```
Comment (adiroiban) [inline paver.sh]: this is already defined below... in the ticket, i just put it here for reference 
please remove it and move the CLEAN_PYTHON_BINARY_DIST_CACHE in the same block with other variable initializations

File: paver.sh [hunk 2]
```diff
@@ -370,79 +380,90 @@ detect_os() {
...
+        get_os_version_digits
+
+        if [ ${OS_VERSION_NUMBERS[1]} -lt 4 ]; then
+            echo "OS X version is too old: ${OS_VERSION_RAW}."
             exit 13
-        else
-            OS="osx"$(echo $osx_version | cut -d'.' -f 1-2 | sed s/\\.//g)
         fi
 
+        OS="osx${OS_VERSION_NUMBERS[0]}${OS_VERSION_NUMBERS[1]}"
```
Comment (adiroiban) [inline paver.sh]: this line should removed as it is not used

File: paver.sh [hunk 3]
```diff
@@ -335,34 +337,42 @@ check_source_folder() {
 }
 
 
+get_os_version_digits() {
+    # OS_VERSION_RAW must be a string composed of integers delimited with dots.
+    # Using '.' as a delimiter, we populate the OS_VERSION_NUMBERS array.
+    IFS=. read -a OS_VERSION_NUMBERS <<< "$OS_VERSION_RAW"
```
Comment (adiroiban) [inline paver.sh]: I don't like global variables, why not pass $OS_VERSION_RAW as input argument for this function?

**Comment by dumol** [comment]
**needs-review**

**Comment by adiroiban** [comment]
I know that you have worked hard on this  so I am accepting it.

50 lines of codes of which 30% are comments is  a smell of unmaintainable code

For aix we only support aix 5.3 and 7.1... and for solaris only solaris 10... I still think that using a case switch makes the code simpler.

But I also hate Bash (and shell scripting)  so maybe is just me who don't like the shell code :)

Thanks!

approved-at  4a9a209

**Comment by adiroiban** [comment]
I also hate code without tests ... and the code from paver.sh has zero tests :(

**Comment by dumol** [comment]
> I know that you have worked hard on this so I am accepting it.

I'm not at all happy with such a reason. Changes should be accepted on technical reasons, not in relation to how long it took and/or how hard it was to write them. And this wasn't much hard work anyway.

> 50 lines of codes of which 30% are comments is a smell of unmaintainable code

If the new code looks unmaintainable we should make it maintainable. Do you want me to dumb down the new `check_os_version` function? It would help if you are willing to accept one more global variable in this script. It already has quite a few already… In the same vein, I still think comparing strings is the most straightforward solution here, with a bit of care for special cases like Solaris and OS X. You can see in the changes that, for SunOS, even when comparing integers, I do that **already**, because we really do not care for the major version. I can do an exception for Darwin too and things would be much simpler. These two changes will get rid of the need for advanced Bash tricks in the new function: IFS and arrays to only compare integers and using eval to return fancy values.


>For aix we only support aix 5.3 and 7.1... and for solaris only solaris 10... I still think that using a case switch makes the code simpler.

Our documentation says we support AIX 6.1 and Solaris 11. The fact that we currently only test AIX 5.3/7.1 and Solaris 10 shouldn't be reflected in code that is published for general use. When the logic is that we support AIX 5.3 and newer and Solaris 10 and newer, the code should reflect that. Bash scripting is expressive enough to allow it. And if we wouldn't be so anal about global variables and string comparisons in a shell script we could even do it in a simple way… :)

Anyway, I went full steam and removed the last `case $var in` in `detect_os()` for consistency. If you think the whole approach is wrong, let's discuss this and find the optimal solution. Thanks!

**needs-review**

**Comment by adiroiban** [comment]
As long as we don't have a buildslave for an OS we are lying in our documentation and we should fix it.
The documentation should state that is might work on other platforms, but they should contact us.

The download page is more accurate.

---------

Beside technical reason we also have a budget and we should spend the time/money as best as we can.
I don't want to invest to much into paver.sh script so as long as it works I am ok with it.

paver.sh is not used in production so we should be fine if it is not perfect... we can slowly improve it.

Thanks!

approved-at  3038aba

**Pull request closed without merging**